pyro/util.py

```diff
@@ -365,10 +384,10 @@ def check_site_shape(site, max_iarange_nesting):
 
     # Check for incorrect iarange placement on the right of max_iarange_nesting.
     for actual_size, expected_size in zip_longest(reversed(actual_shape), reversed(expected_shape), fillvalue=1):
-        if expected_size != -1 and actual_size not in (1, expected_size):
+        if expected_size != -1 and expected_size != actual_size:
```

Comment:
This one line motivates the entire PR:
```diff
- if ... and actual_size not in (1, expected_size)  # allow broadcasting
+ if ... and actual_size != expected_size           # disallow broadcasting
```
With this new stricter requirement, we can more precisely track each `sample` statement's `iarange`s, and thereby support correct scaling (from subsampling) and parallel enumeration.